### PR TITLE
Security

### LINKED ISSUE #13: Security
Review security concerns of the application such as but not limited to:
- DDoS vulnerabilities
- Injection

### PR DESCRIPTION
**Pull request by Jordan141**
### This PR will handle the updates to the backend to popular malicious attacks will not work against the server such as:

- [x] Brute force protection
- [x] Captcha all registration and login requests, maybe article creation
- [x] Use Anti-CSRF tokens
- [x] Prevent HTTP Parameter Pollution
- [x] Sanitise returned objects (such as users) to prevent unintended sensitive/personal information disclosure
- ~~Ensure errors are handled and logged. uncaughtExceptions and Error emitters.~~ #10 
- [x] Secure cookies by setting appropriate cookie flags.
- [x] Implement Strict Transport Layer Security
- [x] X-Frame options
- [x] X-Content-Type-Options
- [x] Content Security Policy
-  ~~Control pages that are allowed to be cached/pragma.~~ Lose too much performance for the small benefits
- [x] X-Download-Options
- ~~Expect Certificate Transparency~~ SSL related
- ~~Public-Key pins (Maybe, must research)~~ SSL related
- [x] Remove X-Powered-By Header
- [x] Implement rate-limiter
- [x] Input validation and sanitiser
- ~~Implement Winston and implement activity logging. Possibly another PR~~ #10 
- ~~Implement SSL~~

Resolves #13 , #8

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -39,28 +43,48 @@ mongoose.connect(`mongodb://mongo_db:27017/${db.name}`,
...
+app.use((req, res, next) => {
+    if(tooBusy()) {
+        return res.sendStatus(503)
+    }
+    next()
+})
+
+//Bruteforce prevention
+const apiLimiter = rateLimit({
+    windowMs: 15 * 60 * 1000, //15 minutes
```
Comment (michaelKurowski) [inline app.js:64]: One request per 15 minutes?

File: app.js [hunk 2]
```diff
@@ -39,28 +43,48 @@ mongoose.connect(`mongodb://mongo_db:27017/${db.name}`,
...
+
 //Pass Redis connection to middleware for easy access 
 app.use((req, res, next) => {
     req.redis = redisClient
     next()
 })
 
 //Express setup
-app.use(bodyParser.urlencoded({extended: true}))
+app.use(bodyParser.urlencoded({extended: true, limit: '1kb'}))
```
Comment (michaelKurowski) [inline app.js:77]: mind extracting it to the top of the file as constant?

File: app.js [hunk 3]
```diff
@@ -69,6 +93,24 @@ app.use(methodOverride('_method'))
...
 
+//Set up HTTP Parameter Pollution
+app.use(require('hpp')())
+
+//Set up Helmet
+//app.use(helmet.hsts()) //HTTP Strict-Transport-Security enable once ssl
+app.use(helmet.frameguard())
+app.use(helmet.noSniff())
+app.use(helmet.ieNoOpen())
+app.use(helmet.hidePoweredBy({setTo: 'Whisky Powered.'}))
```
Comment (michaelKurowski) [inline app.js:104]: We're opensource tho

Comment (michaelKurowski) [reply app.js:104]: or is it private?

File: routes/index.js [hunk 4]
```diff
@@ -3,18 +3,32 @@ const router = express.Router()
...
+const {isLoggedIn, checkCaptcha} = require('../middleware')
+const validator = require('validator')
+const svgCaptcha = require('svg-captcha')
+const csrf = require('csurf')
+const rateLimiter = require('express-rate-limit')
+
+const csrfProtection = csrf({ cookie: true });
+
+const authLimit = rateLimiter({
+    windowMs: 60 * 60 * 1000,
```
Comment (michaelKurowski) [inline routes/index.js:15]: Is this really an hour? Looks like 60h

File: app.js (hunk 1 above)
Comment (Jordan141) [reply app.js:63]: https://www.npmjs.com/package/express-rate-limit

I've set it to 100 requests per 15 minutes

File: app.js (hunk 2 above)
Comment (Jordan141) [reply app.js:76]: Sure

File: app.js (hunk 3 above)
Comment (Jordan141) [reply app.js:103]: It's true that we're open source, but anyone trying to utilize that header technique to gain information about our stack would be malicious. It provides some minimal security with no cons to user experience.

File: routes/index.js (hunk 4 above)
Comment (Jordan141) [reply routes/index.js:15]: 1000ms times 60 is one minute, times 60 again for an hour

**Review by michaelKurowski — APPROVED** [review]

**Pull request merged**